src/hotspot/cpu/riscv/assembler_riscv.hpp

```diff
@@ -928,209 +890,408 @@ enum operand_size { int8, int16, int32, uint32, int64 };
-    patch((address)&insn, 14, 12, rm);                                  \
-    patch((address)&insn, 24, 20, funct5);                              \
-    patch((address)&insn, 31, 25, funct7);                              \
-    patch_reg((address)&insn, 7, Rd);                                   \
-    patch_reg((address)&insn, 15, Rs1);                                 \
-    emit(insn);                                                         \
+  template <FmtPrecision Fmt, uint8_t funct5>
+  void fp_base(FloatRegister Rd, Register Rs1, uint8_t Rs2, RoundingMode rm) {
+    guarantee(is_uimm5(Rs2), "Rs2 is out of validity");
+    fp_base<Fmt, funct5>(Rd->raw_encoding(), Rs1->raw_encoding(), Rs2, rm);
   }
 
-  INSN(fcvt_w_s,   0b1010011, 0b00000, 0b1100000);
-  INSN(fcvt_l_s,   0b1010011, 0b00010, 0b1100000);
-  INSN(fcvt_wu_s,  0b1010011, 0b00001, 0b1100000);
-  INSN(fcvt_lu_s,  0b1010011, 0b00011, 0b1100000);
-  INSN(fcvt_w_d,   0b1010011, 0b00000, 0b1100001);
-  INSN(fcvt_wu_d,  0b1010011, 0b00001, 0b1100001);
-  INSN(fcvt_l_d,   0b1010011, 0b00010, 0b1100001);
-  INSN(fcvt_lu_d,  0b1010011, 0b00011, 0b1100001);
-
-#undef INSN
-
-// Float and Double Move Instruction
-#define INSN(NAME, op, funct3, funct5, funct7)       \
-  void NAME(FloatRegister Rd, Register Rs1) {        \
-    unsigned insn = 0;                               \
-    patch((address)&insn, 6, 0, op);                 \
-    patch((address)&insn, 14, 12, funct3);           \
-    patch((address)&insn, 20, funct5);               \
-    patch((address)&insn, 31, 25, funct7);           \
-    patch_reg((address)&insn, 7, Rd);                \
-    patch_reg((address)&insn, 15, Rs1);              \
-    emit(insn);                                      \
-  }
-
-  INSN(fmv_h_x,  0b1010011, 0b000, 0b00000, 0b1111010);
-  INSN(fmv_w_x,  0b1010011, 0b000, 0b00000, 0b1111000);
-  INSN(fmv_d_x,  0b1010011, 0b000, 0b00000, 0b1111001);
-
-#undef INSN
-
-enum fclass_mask {
-  minf       = 1 << 0,   // negative infinite
-  mnorm      = 1 << 1,   // negative normal number
-  msubnorm   = 1 << 2,   // negative subnormal number
-  mzero      = 1 << 3,   // negative zero
-  pzero      = 1 << 4,   // positive zero
-  psubnorm   = 1 << 5,   // positive subnormal number
-  pnorm      = 1 << 6,   // positive normal number
-  pinf       = 1 << 7,   // positive infinite
-  snan       = 1 << 8,   // signaling NaN
-  qnan       = 1 << 9,   // quiet NaN
-  zero       = mzero    | pzero,
-  subnorm    = msubnorm | psubnorm,
-  norm       = mnorm    | pnorm,
-  inf        = minf     | pinf,
-  nan        = snan     | qnan,
-  finite     = zero     | subnorm   | norm,
-};
+  template <FmtPrecision Fmt, uint8_t funct5>
+  void fp_base(FloatRegister Rd, Register Rs1, uint8_t Rs2, int8_t rm) {
+    guarantee(is_uimm5(Rs2), "Rs2 is out of validity");
+    fp_base<Fmt, funct5>(Rd->raw_encoding(), Rs1->raw_encoding(), Rs2, (RoundingMode)rm);
+  }
+
+ public:
+
+  enum FClassBit {
```

Comment:
Since we have composed ones like `zero`, `subnorm` etc. , maybe we should rename this to `FClassBits`?

Reply:
Yes!